workflow/controller/hooks.go

```diff
@@ -0,0 +1,109 @@
+package controller
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/argoproj/argo-workflows/v3/errors"
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+	"github.com/argoproj/argo-workflows/v3/util/template"
+	"github.com/argoproj/argo-workflows/v3/workflow/common"
+	"github.com/argoproj/argo-workflows/v3/workflow/templateresolution"
+)
+
+func IsExitHook(hookName wfv1.LifecycleEvent) bool {
+	return hookName == wfv1.ExitLifecycleEvent
+}
+
+func (woc *wfOperationCtx) executeWfLifeCycleHook(ctx context.Context, tmplCtx *templateresolution.Context) error {
+	if woc.wf.Spec.Hooks != nil && len(woc.wf.Spec.Hooks) > 0 {
```

Comment:
minor - len(nil) == 0, you don't need nil check on this line,

moreover, as you're looping on 20, you don't need this guard at all `range nil` does not panic